bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/client/oauth2/OAuthError.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.auth.client.oauth2;
+
+/**
+ * This is an exception class for OAUTH specific errors. To keep it simple,
+ * this exception class serves for both Authorization Request and Authorization Grant
+ * error response
+ *
+ * {@see https://tools.ietf.org/html/rfc6749#section-4.1.2.1}
+ * {@see https://tools.ietf.org/html/rfc6749#section-5.2}
+ *
+ * Do not confuse this with Java errors.
```

Comment:
It can easily be confused as this error is used in throws clauses and hence it is considered as java errors but in the context of OAuth feature. This is not an application specific error that you probably wanted to incorporate. I would prefer to have an enumeration listing all the OAuth specific errors and set it in a specific exception or preferably OAuthException. You need to introduce a constructor that accepts such enum as an argument.

```java
public enum OAuthError {
  DUPLICATE_TOKEN(0, "A duplicate token has been provided."),
  INVALID_TOKEN(1, "The token is not valid.");

  private final int code;
  private final String description;

  private Error(int code, String description) {
    this.code = code;
    this.description = description;
  }

  public String getDescription() {
     return description;
  }

  public int getCode() {
     return code;
  }

  @Override
  public String toString() {
    return code + ": " + description;
  }
}
```